url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/RecEditLibrary.java

```diff
@@ -0,0 +1,340 @@
+    {
+        if(aPropNames.size() != aPropVals.size()) throw new CommandException("Nr. of property names and values is different.");
+
+        String[] lPropNamesArr = new String[aPropNames.size()];
+        Object[] lPropValArr = new Object[aPropVals.size()];
+
+        for(int i = 0; i < lPropNamesArr.length; i++) lPropNamesArr[i] = (String) aPropNames.get(i);
+        for(int i = 0; i < lPropValArr.length; i++) lPropValArr[i] = aPropVals.get(i);
+
+        PropertyPage lPage = new PropertyPage(lPropNamesArr, lPropValArr);
+        int result = JOptionPane.showOptionDialog(null, lPage, "Property Editor",JOptionPane.OK_CANCEL_OPTION ,JOptionPane.PLAIN_MESSAGE, null, null, null);
+        return result==JOptionPane.OK_OPTION?lPage.getValues():null;
+    }
+}
+
+class RowEditorModel
+{
+    @SuppressWarnings("unchecked")
+    private Hashtable data;
+
+    public RowEditorModel()
+    {
+        data = new Hashtable();
+    }
+
+    @SuppressWarnings("unchecked")
+    public void addEditorForRow(int row, TableCellEditor e)
+    {
+        data.put(new Integer(row), e);
+    }
+
+    public void removeEditorForRow(int row)
+    {
+        data.remove(new Integer(row));
+    }
+
+    public TableCellEditor getEditor(int row)
+    {
+        return (TableCellEditor) data.get(new Integer(row));
+    }
+}
+
+@SuppressWarnings("serial")
+class JPropTable extends JTable
+{
+    protected RowEditorModel rm;
+
+    public JPropTable()
+    {
+        super();
+        rm = null;
+    }
+
+    public JPropTable(TableModel tm)
+    {
+        super(tm);
+        rm = null;
+    }
+
+    public JPropTable(TableModel tm, TableColumnModel cm)
+    {
+        super(tm,cm);
+        rm = null;
+    }
+
+    public JPropTable(TableModel tm, TableColumnModel cm,
+                      ListSelectionModel sm)
+    {
+        super(tm,cm,sm);
+        rm = null;
+    }
+
+    public JPropTable(int rows, int cols)
+    {
+        super(rows,cols);
+        rm = null;
+    }
+
+    public JPropTable(final Vector rowData, final Vector columnNames)
+    {
+        super(rowData, columnNames);
+        rm = null;
+    }
+
+    public JPropTable(final Object[][] rowData, final Object[] colNames)
+    {
+        super(rowData, colNames);
+        rm = null;
+    }
+
+    // new constructor
+    public JPropTable(TableModel tm, RowEditorModel rm)
+    {
+        super(tm,null,null);
+        this.rm = rm;
+    }
+
+    public void setRowEditorModel(RowEditorModel rm)
+    {
+        this.rm = rm;
+    }
+
+    public RowEditorModel getRowEditorModel()
+    {
+        return rm;
+    }
+
+    public TableCellEditor getCellEditor(int row, int col)
+    {
+        if (rm != null)
+        {
+            TableCellEditor tmpEditor = rm.getEditor(row);
+            if (tmpEditor != null) return tmpEditor;
+        }
+        return super.getCellEditor(row,col);
+    }
+
+    public void changeSelection(int row, int col, boolean toggle, boolean expand)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.